src/charm.py

```diff
@@ -84,100 +79,101 @@ def _on_config_changed(self, event):
         self._configure(event)
 
     def _configure(self, event):
-        restart = False
-
         if not self._container.can_connect():
             self.unit.status = WaitingStatus("Waiting for Pebble ready")
             return
 
         try:
-            restart = self._set_alerts(self._container)
-        except PebbleError as e:
-            logger.error("Failed to push updated alert files: %s", e)
-            self.unit.status = BlockedStatus("Failed to push updated alert files; see debug logs")
+            self._set_alerts()
+            self._set_mimir_config()
+            self._set_pebble_layer()
+        except BlockedStatusError as e:
+            self.unit.status = BlockedStatus(e.message)
             return
 
+        if self._restart:
+            self._container.restart(self._name)
+            logger.info("Mimir (re)started")
+
+        self.remote_write_provider.update_endpoint()
+        self.unit.status = ActiveStatus()
+
+    def _set_mimir_version(self) -> bool:
+        version = self._mimir_version
+
+        if version is None:
+            logger.debug("Cannot set workload version at this time: could not get Mimir version.")
+            return False
+
+        self.unit.set_workload_version(version)
+        return True
+
+    def _set_pebble_layer(self):
         current_layer = self._container.get_plan().to_dict()
         new_layer = self._pebble_layer
 
         if "services" not in current_layer or DeepDiff(
             current_layer["services"], new_layer["services"], ignore_order=True
         ):
-            restart = True
+            self._container.add_layer(self._name, new_layer, combine=True)
+            self._restart = True
 
+    def _set_mimir_config(self):
         config = self._mimir_config
 
         try:
             if self._current_mimir_config != config:
                 config_as_yaml = yaml.safe_dump(config)
                 self._container.push(MIMIR_CONFIG, config_as_yaml, make_dirs=True)
                 logger.info("Pushed new Mimir configuration")
-                restart = True
+                self._restart = True
         except (ProtocolError, PathError) as e:
```

Comment:
- Do we need `PathError` if we have `make_dirs=True`?
- I'm happy with the try block but I thought we said we prefer to have the "pretty but partial" `can_connect` instead of the "rigorous but ugly" `except`s.
- Should we have all `push`s inside try block?

Reply:
Yes, seems we can avoid `PathError`...

The `_set_mimir_config()` method is called inside `_configure()` method **after** a `.can_connect()`:

```python
    def _configure(self, event):
        if not self._container.can_connect():
            self.unit.status = WaitingStatus("Waiting for Pebble ready")
            return

        try:
            self._set_alerts()
            self._set_mimir_config()
            self._set_pebble_layer()
        ...
```
The idea behind the `try...excpet` block is to eventually raise a `BlockedStatusError` exception